contracts/cw1-whitelist-ng/src/contract.rs

```diff
@@ -0,0 +1,440 @@
+            admins: vec![alice.to_owned(), bob.to_owned()],
+            mutable: false,
+        };
+        assert_eq!(
+            contract.admin_list(deps.as_ref(), mock_env()).unwrap(),
+            expected
+        );
+
+        // and now alice cannot change it again
+        let info = mock_info(&alice, &[]);
+        let err = contract
+            .update_admins(deps.as_mut(), mock_env(), info, vec![alice.to_owned()])
+            .unwrap_err();
+        assert_eq!(err, ContractError::Unauthorized {});
+    }
+
+    #[test]
+    fn execute_messages_has_proper_permissions() {
+        let contract = Cw1WhitelistContract::native();
+        let mut deps = mock_dependencies(&[]);
+
+        let alice = "alice";
+        let bob = "bob";
+        let carl = "carl";
+
+        // instantiate the contract
+        let admins = vec![alice.to_owned(), carl.to_owned()];
+        let info = mock_info(&bob, &[]);
+        contract
+            .instantiate(deps.as_mut(), mock_env(), info, admins, false)
+            .unwrap();
+
+        let freeze = WhitelistExecMsg::Freeze {};
+        let msgs = vec![
+            BankMsg::Send {
+                to_address: bob.to_string(),
+                amount: coins(10000, "DAI"),
+            }
+            .into(),
+            WasmMsg::Execute {
+                contract_addr: "some contract".into(),
+                msg: to_binary(&freeze).unwrap(),
+                funds: vec![],
+            }
+            .into(),
+        ];
+
+        // bob cannot execute them
+        let info = mock_info(&bob, &[]);
+        let err = contract
+            .execute(deps.as_mut(), mock_env(), info, msgs.clone())
+            .unwrap_err();
+        assert_eq!(err, ContractError::Unauthorized {});
+
+        // but carl can
+        let info = mock_info(&carl, &[]);
+        let res = contract
+            .execute(deps.as_mut(), mock_env(), info, msgs.clone())
+            .unwrap();
+        assert_eq!(
+            res.messages,
+            msgs.into_iter().map(SubMsg::new).collect::<Vec<_>>()
+        );
+        assert_eq!(res.attributes, [("action", "execute")]);
+    }
+
+    #[test]
+    fn execute_custom_messages_works() {
+        let contract = Cw1WhitelistContract::<String>::new();
+        let mut deps = mock_dependencies(&[]);
+        let alice = "alice";
+
+        let admins = vec![alice.to_owned()];
+        let info = mock_info(&alice, &[]);
+        contract
+            .instantiate(deps.as_mut(), mock_env(), info, admins, false)
+            .unwrap();
+
+        let msgs = vec![CosmosMsg::Custom("msg".to_owned())];
+
+        let res = contract
+            .execute(
+                deps.as_mut(),
+                mock_env(),
+                mock_info(&alice, &[]),
+                msgs.clone(),
+            )
+            .unwrap();
+
+        assert_eq!(
+            res.messages,
+            msgs.into_iter().map(SubMsg::new).collect::<Vec<_>>()
```

Comment:
This chain could make a nice helper function (in cosmwasm-std or cw-plus).
Or not... just a thought